Presentational Roles Conflict Resolution - global WAI-ARIA states or properties on inherited presentational roles

giacomo-petri opened the pull request:
🚀 **Netlify Preview**:
🔄 **this PR updates the following sspecs**:
- [ARIA preview](https://deploy-preview-2237--wai-aria.netlify.app/index.html) &mdash; [ARIA diff](https://services.w3.org/htmldiff?doc1=https%3A%2F%2Fw3c.github.io%2Faria%2F&doc2=https%3A%2F%2Fdeploy-preview-2237--wai-aria.netlify.app%2Findex.html)

Closes: #2233

- clarified global WAI-ARIA states or properties on inherited presentational roles
- reorganized the structure of the 3rd bullet to visually and semantically emphasize that the examples below strictly refer to this bullet. The final paragraph related to the author's responsibility is instead applicable to all points.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/aria/pull/2237.html" title="Last updated on Dec 12, 2025, 11:07 AM UTC (74750ac)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/aria/2237/bf8e158...74750ac.html" title="Last updated on Dec 12, 2025, 11:07 AM UTC (74750ac)">Diff</a>

giacomo-petri committed:
clarified global WAI-ARIA states or properties on inherited presentational roles

- clarified global WAI-ARIA states or properties on inherited presentational roles
- reorganized the structure of the 3rd bullet to visually and semantically emphasize that the examples below strictly refer to this bullet. The final paragraph related to the author's responsibility is instead applicable to all points.

scottaohara commented on index.html:
```diff
@@ -13524,16 +13524,18 @@ <h3>Presentational Roles Conflict Resolution</h3>
 			<li>If an <a href="#mustContain">allowed child element</a> has an explicit non-presentational role, user agents MUST ignore an inherited presentational role and expose the element with its explicit role. If the action of exposing the explicit role causes the accessibility tree to be malformed, the expected results are undefined.</li>
-      <li>If an element has <a>global</a> WAI-ARIA states or properties, user agents MUST ignore the <rref>none</rref>/<rref>presentation</rref> role and instead expose the element's implicit role. However, if an element has only non-global, role-specific WAI-ARIA states or properties, the element MUST NOT be exposed unless the presentational role is inherited and an explicit non-presentational r ...
+			<li>
+				<p>If an element with an explicit presentational role has <a>global</a> WAI-ARIA states or properties, user agents MUST ignore the <rref>none</rref>/<rref>presentation</rref> role and instead expose the element's implicit role. Conversely, if the presentational role is inherited and the element has <a>global</a> WAI-ARIA states or properties, user agents MUST NOT expose the element, maintaini ...
```
>the element MUST NOT be exposed unless the

the element or the element's role?

giacomo-petri replied:
Good point, Scott!
While looking through the ARIA specifications, I found similar wording and re-used it, assuming the intent was to not expose the element's role rather than the element itself (I can't recall the exact location right now). This interpretation seems supported by the following statement:

> Thus, the [none](https://w3c.github.io/aria/#none)/[presentation](https://w3c.github.io/aria/#presentation) role causes a given element to be treated as having no role or to be removed from the [accessibility tree](https://w3c.github.io/aria/#dfn-accessibility-tree), but does not cause the content contained within the element to be removed from the accessibility tree.

But I'm more than happy to use your wording to clarify that we do not expect the content to be ignored, only the element's role.

giacomo-petri replied:
Oh, I found it! It was the original statement I didn't change:

> However, if an element has only non-global, role-specific WAI-ARIA states or properties, the element MUST NOT be exposed unless the presentational role is inherited and an explicit non-presentational role is applied.

I'll update both

giacomo-petri commented on index.html:
```diff
@@ -13524,16 +13524,18 @@ <h3>Presentational Roles Conflict Resolution</h3>
 			<li>If an <a href="#mustContain">allowed child element</a> has an explicit non-presentational role, user agents MUST ignore an inherited presentational role and expose the element with its explicit role. If the action of exposing the explicit role causes the accessibility tree to be malformed, the expected results are undefined.</li>
-      <li>If an element has <a>global</a> WAI-ARIA states or properties, user agents MUST ignore the <rref>none</rref>/<rref>presentation</rref> role and instead expose the element's implicit role. However, if an element has only non-global, role-specific WAI-ARIA states or properties, the element MUST NOT be exposed unless the presentational role is inherited and an explicit non-presentational r ...
+			<li>
+				<p>If an element with an explicit presentational role has <a>global</a> WAI-ARIA states or properties, user agents MUST ignore the <rref>none</rref>/<rref>presentation</rref> role and instead expose the element's implicit role. Conversely, if the presentational role is inherited and the element has <a>global</a> WAI-ARIA states or properties, user agents MUST NOT expose the element, maintaini ...
```
```suggestion
				<p>If an element with an explicit presentational role has <a>global</a> WAI-ARIA states or properties, user agents MUST ignore the <rref>none</rref>/<rref>presentation</rref> role and instead expose the element's implicit role. Conversely, if the presentational role is inherited and the element has <a>global</a> WAI-ARIA states or properties, user agents MUST NOT expose the element's role, maintaining its presentational role regardless of any <a>global</a> ARIA state or property. However, if an element has only non-global, role-specific WAI-ARIA states or properties, the element's role MUST NOT be exposed unless the presentational role is inherited and an explicit non-presentational role is applied.</p>
```

giacomo-petri committed:
Update index.html

pkra commented:
@giacomo-petri I believe you need to rebase or cherry pick the commit from #2244

giacomo-petri commented:
@pkra, it seems there are other stray end tags...

cookiecrook commented:
Unassigning myself because I don't have the capacity to review a 30,000+ line diff.

jnurthen commented:
> Unassigning myself because I don't have the capacity to review a 30,000+ line diff. It won't render on GitHub because it's too large.

Hmm yeah - We could probably review this by just looking at the commits (other than the prettier one). But it may be easier to just cherry pick them into a new PR. @spectranaut @pkra do you have any suggestions?

pkra commented:
@jnurthen not sure what's going on here. The prettier action should ignore the aria  spec but it seems to have formatted it. 

We can try reverting the prettier commit. But if the actions are flaky, I'm wondering if it isn't time to bite the bullet and activate prettier (and update all PRs). Since I did that manually for the PRs from other specs, I should be able to do this for ARIA.

pkra commented:
I reversed the prettier commit, rebased on main, and force-pushed here. It looks like this solved the prettier problem.

@giacomo-petri, @jnurthen, @scottaohara, @cookiecrook could please take another look?

pkra commented:
@giacomo-petri I've rebased on main.

pkra committed:
Merge branch 'main' into giacomo-petri-presentational-roles-conflict-resolution-global-states-properties

spectranaut commented:
@giacomo-petri does this need tests?

jnurthen reviewed (changes requested):
I think this is clearer.

giacomo-petri committed:
Update index.html

Co-authored-by: James Nurthen <jnurthen@users.noreply.github.com>

giacomo-petri commented on index.html:
```diff
@@ -16680,25 +16680,18 @@ <h3>Presentational Roles Conflict Resolution</h3>
-            If an element has <a>global</a> WAI-ARIA states or properties, user agents MUST ignore the <rref>none</rref>/<rref>presentation</rref> role and instead expose the element's implicit role.
-            However, if an element has only non-global, role-specific WAI-ARIA states or properties, the element MUST NOT be exposed unless the presentational role is inherited and an explicit
-            non-presentational role is applied.
+            <p>If an element with an explicit presentational role has <a>global</a> WAI-ARIA states or properties, user agents MUST ignore the <rref>none</rref>/<rref>presentation</rref> role and instead expose the element's implicit role. Conversely, if the presentational role is inherited and the element has <a>global</a> WAI-ARIA states or properties, user agents MUST NOT expose the element's  ...
```
Thanks @jnurthen, great improvement.

giacomo-petri commented on index.html:
```diff
@@ -16680,25 +16680,28 @@ <h3>Presentational Roles Conflict Resolution</h3>
+              &lt;h1 role="none" aria-level="2"&gt; Sample Content &lt;/h1&gt;
+            </pre>
+          </li>
           </li>
```
```suggestion
```

giacomo-petri committed:
Update index.html

front-endian commented on index.html:
```diff
@@ -16680,25 +16680,27 @@ <h3>Presentational Roles Conflict Resolution</h3>
             explicit role. If the action of exposing the explicit role causes the accessibility tree to be malformed, the expected results are undefined.
```
Happy to move this to a separate issue to avoid bloating this PR, but does the following only apply to this second exception, or all of them?
> If the action of exposing the explicit role causes the accessibility tree to be malformed, the expected results are undefined.

front-endian replied:
Pulled this out into a separate issue so this PR can proceed faster: #2638

front-endian commented on index.html:
```diff
@@ -16680,25 +16680,27 @@ <h3>Presentational Roles Conflict Resolution</h3>
+                  <strong>Inherited presentational role:</strong> User agents MUST NOT expose the element's role, maintaining its presentational role regardless of any <a>global</a> ARIA state or property.
+                </li>
+              </ul>
+              <p>However, if an element has only non-global, role-specific WAI-ARIA states or properties, the element's role MUST NOT be exposed unless the presentational role is inherited and an explicit non-presentational role is applied.</p>
```
It is a bit hard to parse all this in context since there is some jumping back and forth between exceptions and non-exceptions.

Because all of this is preempted by,

> User agents MUST NOT expose elements having explicit or inherited presentational role in the accessibility tree, with these exceptions

... I wonder if this might be easier to parse if some of the parts explaining when elements _must not_ be exposed were left out? That way the reader doesn't have to jump back and forth between examples of "must not" and "must" and can focus in on the few exceptions.

Maybe something like:

> *[first two items in list of exceptions]*
> - If an element has global WAI-ARIA states or properties, and has an explicit presentational role, user agents MUST ignore the none/presentation role and instead expose the element's implicit role.
> - If an element has an inherited presentational role but an explicit non-presentational role is applied to it as well, user agents MUST ignore the inherited presentational role and expose the explicit role applied to the element.

Then maybe the, "however" isn't needed because the above exception clearly states it only applies when there are **global** states or properties.

front-endian commented on index.html:
```diff
@@ -16680,25 +16680,27 @@ <h3>Presentational Roles Conflict Resolution</h3>
+              <p>However, if an element has only non-global, role-specific WAI-ARIA states or properties, the element's role MUST NOT be exposed unless the presentational role is inherited and an explicit non-presentational role is applied.</p>
+
+            <p class="note">Some <a>global</a> WAI-ARIA states and properties are <a>prohibited</a> on certain roles. These states and properties are still considered global for the purposes of Presentational Role Conflict resolution.</p>
+            <p>For example, <pref>aria-describedby</pref> is a global attribute and would always be applied; <pref>aria-level</pref> is not a global attribute and would therefore only apply if the element was not in a presentational state.</p>
```
Small edit to keep tenses consistent:

```suggestion
            <p>For example, <pref>aria-describedby</pref> is a global attribute and would always be applied; <pref>aria-level</pref> is not a global attribute and would therefore only apply if the element is not in a presentational state.</p>
```

front-endian reviewed (changes requested):
Have some requested changes to simplify/clarify a few parts.

giacomo-petri committed:
Update index.html

Co-authored-by: front-endian <11234702+front-endian@users.noreply.github.com>

front-endian reviewed (commented):
Suggesting a more minor edit to preserve the easier to follow nested list.

giacomo-petri committed:
Update index.html

giacomo-petri committed:
Update index.html

front-endian reviewed (approved):
@giacomo-petri , sorry for the delay re-reviewing this, but Option 2 looks great to me. Clears up all my concerns and is easy to follow. Thank you!

Approved assuming we go with something like Option 2.

cookiecrook commented on index.html:
```diff
@@ -16680,25 +16681,53 @@ <h3>Presentational Roles Conflict Resolution</h3>
+                  <strong>Inherited presentational role:</strong> User agents MUST NOT expose the element's role, maintaining its presentational role regardless of any <a>global</a> ARIA state or property.
+                </li>
+              </ul>
+              <p>However, if an element has only non-global, role-specific WAI-ARIA states or properties, the element's role MUST NOT be exposed unless the presentational role is inherited and an explicit non-presentational role is applied.</p>
```
FYI @front-endian since she raised a similar clarity concern in the resolved comment before this one.

front-endian replied:
This looked like a good edit to me.

> How can there be both an inherited presentational role and an explicit non-presentational role?

I think the second half is trying to clarify that an explicit `role="blah"` overrides any situation where a presentational role would be implicitly applied? It does feel like over clarifying things and I'm not sure I understand how this situation would happen on a website.

cookiecrook replied:
> However, if an element with an explicit or implicit presentational role has only non-global, role-specific WAI-ARIA attributes, user agents MUST NOT expose the element's role.

This part is now ambiguous too I think because the earlier part of the sentence referenced two different potential roles: "MUST NOT expose the element's role."

UAs MUST NOT expose the `none` role? Or UAs MUST NOT the other role that's associated with the role-specific attributes?

cookiecrook commented on index.html:
```diff
@@ -16680,25 +16681,53 @@ <h3>Presentational Roles Conflict Resolution</h3>
+		<h4>Option 2</h4>
+		<ul>
+          <li>
+            If an element is focusable, user agents MUST ignore the <rref>none</rref>/<rref>presentation</rref> role and expose the element with its implicit role, in order to ensure that the element
```
Thinking about this more, I don't think we can force engines to expose the implicit role on elements with a validly inherited role of none, like in my list `example` from above.

```html
<ul role="none">
  <tr>
    <td tabindex="0"><!-- is this an orphaned `cell` or a minimum role… maybe `group`? -->
```

Maybe the solution is (hand-wavy, half-considered draft text follows):

- If element is focusable or otherwise has ~some attributes], ignore any **_explicit_** role of none/presentation.
- If element is focusable or otherwise has ~some attributes, and the element element's required accessibility parent has a role of none, do not inherit the none role, and instead, expose the element's minimum role.

This may be difficult to implement, if you all settle on any specific language that can solve this, I'd want to run it by more engine accessibility engineers for each of the major browsers.

cookiecrook commented on index.html:
```diff
@@ -16680,25 +16681,53 @@ <h3>Presentational Roles Conflict Resolution</h3>
+		  <li>
+			If an element with an explicit presentational role has <a>global</a> WAI-ARIA states or properties, user agents MUST ignore the <rref>none</rref>/<rref>presentation</rref> role and instead expose the element’s implicit role.
+		  	<p class="note">Some <a>global</a> WAI-ARIA states and properties are <a>prohibited</a> on certain roles. These states and properties are still considered global for the purposes of Presentational Role Conflict resolution.</p>
+            <p>For example, <pref>aria-describedby</pref> is a global attribute and would always be applied; <pref>aria-level</pref> is not a global attribute and would therefore only apply if the element is not in a presentational state.</p>
```
```suggestion
            <p>For example, <pref>aria-describedby</pref> is a global attribute and would always be applied; <pref>aria-level</pref> is not a global attribute and would therefore only apply if the element does not have an explicit or inherited role of <rref>none</rref> or <rref>presentation</rref>.</p>
```

cookiecrook commented on index.html:
```diff
@@ -16680,25 +16681,53 @@ <h3>Presentational Roles Conflict Resolution</h3>
+            <pre class="example highlight">
+              <span class="comment">&lt;!-- 1. [role="none"] is ignored due to the global aria-describedby property. --&gt;</span>
+              &lt;h1 role="none" aria-describedby="comment-1"&gt; Sample Content &lt;/h1&gt;
+              <span class="comment">&lt;!-- 2. [role="none"] negates both the implicit 'heading' and the non-global aria-level. --&gt;</span>
```
good examples

front-endian commented on index.html:
```diff
@@ -16670,6 +16670,7 @@ <h3>States and Properties</h3>
         <h3>Presentational Roles Conflict Resolution</h3>
         <p>There are a number of ways presentational role conflicts are resolved.</p>
         <p>User agents MUST NOT expose [=elements=] having explicit or inherited presentational role in the accessibility tree, with these exceptions:</p>
```
I second that if we mention what presentation roles do in this section it would be helpful to clarify that given how common/unintuitive/impactful that misunderstanding is.

On the other hand, do we even have to document any of this section since this is starting to duplicated the existing documentation of `role="none"`? Could it be simplified to something along the lines of:
> User agents follow the expected behavior for elements with a presentational role in the accessibility tree, with these exceptions:

giacomo-petri replied:
I've just realised that we don't actually define "Presentational Roles Conflict Resolution" anywhere. We mention that there are several ways such conflicts are resolved, but we never state what it is.

I'll add a definition that explains the concept explicitly, also clarifying that "MUST not expose elements" does not mean "MUST not expose any descendant elements" and maybe including an example to illustrate this.

front-endian commented on index.html:
```diff
@@ -16680,25 +16681,53 @@ <h3>Presentational Roles Conflict Resolution</h3>
+          <li>
+            If an <a href="#mustContain">allowed child element</a> has an explicit non-presentational role, user agents MUST ignore an inherited presentational role and expose the element with its
+            explicit role. If the action of exposing the explicit role causes the accessibility tree to be malformed, the expected results are undefined.
+          </li>
```
@cookiecrook, [this came up previously](https://github.com/w3c/aria/pull/2237#discussion_r2321257351) and there is an open issue to make this change: #2638 . It was moved to a separate issue since this bit is carried over from the current published spec, things were getting a bit scope creep-y with what this PR is intended to fix, and to help this PR make progress.

That being said, I agree that that edit should be made at some point. With #2638 I was waiting until this PR landed before pushing an edit, but I could also push an edit now since this PR is taking a while.

giacomo-petri committed:
Update index.html

Co-authored-by: James Craig <cookiecrook@users.noreply.github.com>

giacomo-petri committed:
Update index.html

Co-authored-by: James Craig <cookiecrook@users.noreply.github.com>

giacomo-petri committed:
Apply suggestions from code review

Co-authored-by: James Craig <cookiecrook@users.noreply.github.com>

giacomo-petri committed:
tentative

giacomo-petri commented:
TODO: Missing condition for non-focusable widget roles(?) nested inside a presentational child element. 

```
<div role="button">
    <button>
```
This scenario is already covered by the focusable bit, where we state that focusable elements must be exposed with their implicit role (need to tweak it based on @cookiecrook [example](https://github.com/w3c/aria/pull/2237#discussion_r2594338802))

----

```
<button>
    <div role="button">
```
However, this case is not covered, because the nested non-focusable button should instead be treated as presentational (i.e., its children should be presentational). Per the ARIA specs, it should be exposed as a generic element, but all user agents currently expose it as a button, as confirmed by the WPT reults in [WPT-50045](https://github.com/web-platform-tests/wpt/pull/50045))

giacomo-petri committed:
tenative

giacomo-petri committed:
Update index.html

removed malformed stuff, which is not related to presentational conflict role resolution

giacomo-petri committed:
Update index.html

giacomo-petri commented:
I've updated the PR:
- I removed option 1 for clarity, since previous comments favored option 2.
- I believe the concerns highlighted in earlier comments are now addressed, with clearer definitions of what happens in scenarios such as:
  ```
  <ul role="none">
    <tr>
      <td tabindex="0">
  ```
    I also added a couple of additional examples to make this behavior clearer.
- While working on:
  ```
  <button>
    <div role="button">button</div>
  </button>
  ```
    I encountered the issue we discussed some time ago regarding "presentational children" and how UAs treat them, which makes the concept of "presentational children" somewhat theoretical in practice. However, by addressing this within the presentational roles conflict resolution, we can clarify that UAs may continue their current behavior for the benefit of end users, while still treating it as a conflict and therefore an authoring error.

giacomo-petri committed:
Merge branch 'main' into giacomo-petri-presentational-roles-conflict-resolution-global-states-properties

giacomo-petri committed:
Update index.html

spectranaut commented on index.html:
```diff
@@ -16741,28 +16741,57 @@ <h3>States and Properties</h3>
+            </pre>
+		  </li>
+		  <li>
+			If an element has a presentational role as a result of <a href="#childrenArePresentational">children presentational</a>, user agents MAY ignore the inherited presentational role and expose the element with its role, to preserve the element's implicit role semantics.
```
It seems to me that this is in agreement with the request to deprecate children presentational, as requested in this issue: https://github.com/w3c/aria/issues/2509

The WG didn't feel comfortable making that change, because there are scenarios where assistive technologies cannot handle that content.

spectranaut reviewed (changes requested):
If we are keeping children presentation (which somehow I've been thrown into question on whether we should) I think we should land this, but without the bullet I commented on, because as far as I can tell that is equivalent to removing the children presentation requirement entirely.
